drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,563 @@
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	/* TODO: check pin against mask */
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		config->flags[pin] &= ~interrupt_bitmask;
+		config->flags[pin] |= GPIO_INT_DISABLE;
+		break;
+
+	case GPIO_INT_MODE_LEVEL:
+	case GPIO_INT_MODE_EDGE:
+		config->flags[pin] &= ~interrupt_bitmask;
+		config->flags[pin] |= (mode | trig);
+		break;
+
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int gpio_sim_manage_callback(struct device *port,
+				    struct gpio_callback *cb, bool set)
+{
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	return gpio_manage_callback(&config->callbacks, cb, set);
+}
+
+static uint32_t gpio_sim_get_pending_int(struct device *dev)
+{
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)dev->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	return config->interrupts;
+}
+
+static const struct gpio_driver_api gpio_sim_driver = {
+	.pin_configure = gpio_sim_pin_configure,
+	.port_get_raw = gpio_sim_port_get_raw,
+	.port_set_masked_raw = gpio_sim_port_set_masked_raw,
+	.port_set_bits_raw = gpio_sim_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_sim_port_clear_bits_raw,
+	.port_toggle_bits = gpio_sim_port_toggle_bits,
+	.pin_interrupt_configure = gpio_sim_pin_interrupt_configure,
+	.manage_callback = gpio_sim_manage_callback,
+	.get_pending_int = gpio_sim_get_pending_int,
+};
+
+static int gpio_sim_init(struct device *dev, size_t n)
+{
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)dev->driver_data;
+	struct gpio_sim_config *config = &gpio_sim_config[n];
+
+	memset(config, 0, sizeof(*config));
+
+	drv_data->config = config;
+	config->common.port_pin_mask =
+		(1ULL << CONFIG_GPIO_SIM_NPINS_PER_PORT) - 1;
+	sys_slist_init(&config->callbacks);
+
+	return k_mutex_init(&config->mu);
+}
+
+/*
+ * Device Initialization
+ */
+
+static int gpio_sim_init_0(struct device *dev)
+{
+	return gpio_sim_init(dev, 0);
+}
+
+DEVICE_AND_API_INIT(gpio_sim_0, DT_INST_LABEL(0), gpio_sim_init_0,
+		    &gpio_sim_data[0], &gpio_sim_config[0], POST_KERNEL,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &gpio_sim_driver);
+
+#if CONFIG_GPIO_SIM_NPORTS >= 2
+static int gpio_sim_init_1(struct device *dev)
+{
+	return gpio_sim_init(dev, 1);
+}
+
+DEVICE_AND_API_INIT(gpio_sim_1, DT_INST_LABEL(1), gpio_sim_init_1,
+		    &gpio_sim_data[1], &gpio_sim_config[1], POST_KERNEL,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &gpio_sim_driver);
+#endif
+
+#if CONFIG_GPIO_SIM_NPORTS >= 3
+static int gpio_sim_init_2(struct device *dev)
+{
+	return gpio_sim_init(dev, 2);
+}
+
+DEVICE_AND_API_INIT(gpio_sim_2, DT_INST_LABEL(2), gpio_sim_init_2,
```

Comment:
If this moves to the DT (i.e. each bank is a separate node) then I hope this code can be simplified.

Reply:
Fixed.